### PR TITLE
mempool: modify `maintain_transaction_pool` to prune txs

### PR DESCRIPTION
**Pull request by tcoratger**
Should close https://github.com/kkrt-labs/kakarot-rpc/issues/1478

- We maintain a hashmap of transactions that were at one time in the pool with the timestamp of the first time we found them.
- We empty this hashmap with the mined transactions that are progressively included in the last block and that are pruned correctly.
- In addition, we also prune transactions that are still in the mempool at the moment we look and that have been in it for more than 5 minutes.
- A unit test is here to validate the proper behaviour.

cc @greged93 @ClementWalter @eugypalu

### CODE REVIEW

File: tests/tests/mempool.rs [hunk 1]
```diff
@@ -351,3 +367,105 @@ pub async fn create_sample_transactions(
...
+
+        // Update the header collection with the padded block number in the database.
+        header_collection
+            .update_one(
+                doc! {"header.number": unpadded_block_number},
+                UpdateModifications::Document(doc! {"$set": {"header.number": padded_block_number}}),
+            )
+            .with_options(UpdateOptions::builder().upsert(true).build())
+            .await
+            .expect("Failed to update block number");
```
Comment (tcoratger) [inline tests/tests/mempool.rs:454]: We need to scope an issue here to be able to add block headers for testing more easily in the database. This behavior is also present in katana. As it is for testing purposes, we do not have a dedicated function in the production flow but we would have to scope an issue to do this properly.

Comment (Eikix) [reply tests/tests/mempool.rs:454]: Can we create an issue and scope it in next tech workshop?

Comment (tcoratger) [reply tests/tests/mempool.rs:454]: done https://github.com/kkrt-labs/kakarot-rpc/issues/1489

File: tests/tests/mempool.rs
Comment (ClementWalter) [inline tests/tests/mempool.rs:1]: can't we divide the duration by 10 to avoid slowing down the tests to much?

Comment (tcoratger) [reply tests/tests/mempool.rs:1]: Fixed, I tried but it seemed that the CI is very sensitive to very short times, let's see if the CI still passes.

I also reduced to 9 iterations in the loop to be sure that the calculation of the 10th does not arrive at the end of the allowed time. I would say that like this we are safe from a flaky test.

File: src/pool/mempool.rs [hunk 2]
```diff
@@ -338,7 +352,31 @@ where
...
+
+                        // Prune mined transactions from the mempool mapping
+                        for tx_hash in &mined_transactions {
+                            mempool_transactions.remove(tx_hash);
+                        }
+
+                        // Prune transactions that have been in the mempool for more than 5 minutes
+                        let now = Instant::now();
+
+                        for (tx_hash, timestamp) in &mempool_transactions.clone() {
```
Comment (greged93) [inline src/pool/mempool.rs:366]: do we need the `clone()`?

File: src/pool/mempool.rs [hunk 3]
```diff
@@ -338,7 +352,31 @@ where
...
+                            // - If the transaction has been in the mempool for more than 5 minutes
+                            // - And the transaction is in the mempool right now
+                            if now.duration_since(*timestamp) > prune_duration && eth_client.mempool().contains(tx_hash)
+                            {
+                                tracing::warn!("Transaction {} in mempool for more than 5 minutes. Pruning.", tx_hash);
```
Comment (greged93) [inline src/pool/mempool.rs:371]:
```suggestion
                                tracing::warn!(target: "maintain_transaction_pool", tx_hash, "pruning");
```

File: src/main.rs [hunk 4]
```diff
@@ -57,8 +56,11 @@ async fn main() -> Result<()> {
         var("RELAYERS_ADDRESSES")?.split(',').filter_map(|addr| Felt::from_str(addr).ok()).collect::<Vec<_>>();
     AccountManager::from_addresses(addresses, Arc::clone(&eth_client)).await?.start();
 
+    // Transactions should be pruned after 5 minutes in the mempool
+    let prune_duration = Duration::from_secs(300);
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/main.rs]: Fixed, actually we already had a constant file there so I just pushed this declaration here

File: src/pool/mempool.rs [hunk 5]
```diff
@@ -268,13 +268,27 @@ where
...
+        // Mapping to store the transactions in the mempool with a timestamp to potentially prune them
+        let mut mempool_transactions = HashMap::new();
+
         loop {
+            // Adding the transactions to the mempool mapping with a timestamp
+            for tx in eth_client
+                .mempool()
+                .queued_transactions()
+                .into_iter()
+                .chain(eth_client.mempool().pending_transactions())
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/pool/mempool.rs:301]: As discussed separately this is not possible to manipulate the transactions like this right now.

File: src/pool/mempool.rs (hunk 2 above)
Comment (greged93) [reply src/pool/mempool.rs:366]: ah ok but then we can just do `mempool_transactions.clone()` no? no need for the reference I believe.

Comment (tcoratger) [reply src/pool/mempool.rs]: yes we can just modified it, this just induced some inverting some "take by value" or "take by reference" in the loop but this is cleaner :)

**Review by greged93 — APPROVED** [review]

**Pull request merged**